### PR TITLE
Add Double Click to Open in Default Editor

### LINKED ISSUE #2620: Double click to open with selected external editor
### Description

The old client I often used to have an overview of the files I worked on in a large project. By quickly switching to GH Desktop and double clicking on the file I was able to jump to the file in the editor. The extra click to open the context menu makes this a lot slower and the client unsuitable for this workflow.

Would ❤️  to see this double-click action being introduced in the new client.

### Version

**GitHub Desktop version:** 0.8.0

### PR DESCRIPTION
**Pull request by digitalmaster**
Closes #2620

## Description

- [x] Add Double Click To Open Changed file In Default Program
- [x] Add same interaction to all changes list views
- [x] [REVERTED] ~~Add same interaction to double click diff line number~~
- [x] Investigate: Bold "Open with Default Program" context menu item (Windows)

### Preview

https://user-images.githubusercontent.com/1279335/233219332-a7469a62-007e-4eee-a720-19ff636c2489.mp4

# Testing Instructions

**Double Clicking Files Lists**
1. Find a change list (Main Change List, History Change List, Stash Change List)
2. Double Click on the file name

**Expectation:** Should open file in default editor

**Double Clicking Diff Line Numbers**
1. Find a diff view  (Main Change List, History Change List, Stash Change List)
2. Double Click on line number

**Expectation:** If vscode `code` command is installed on path, should open file at line number in default editor.

https://user-images.githubusercontent.com/1279335/236348957-315e6305-d376-4102-a48f-92c4e12c0d40.mp4

**Expectation:** Otherwise, it shows this dialog message
CleanShot 2023-05-04 at 15 58 39@2x
**Expectation:** If the default editor doesn't have a supported `editorOpenToLineCommands` function defined, it just opens the file without going to that specific line. 

## Release notes

Notes:

### CODE REVIEW

**Review by tidy-dev — DISMISSED** [review]
👋 Great start! This approach does work. However, it is not consistent with our current approach to interacting with list items. If you look in the `changes-list.tsx` at the `<List>` element. You will see an implementation of `onRowClick` and `onRowContextMenu`. To be consistent, it would be better to implement a `onRowDoubleClick`. This would also allow you to reuse some of the logic in the `getDefaultContextMenu` and `getOpenInExternalEditorMenuItem` for opening in an external editor. (May even be ideal to create a shared method for the action part of the context menu bit so both implementations are ensured to do the same thing)

**Comment by digitalmaster** [comment]
Hi @tidy-dev 👋🏽 - Thanks for taking the time to review. 

Great feedback... Iterating now...

**Comment by digitalmaster** [comment]
Had some time on a plane to work on this yesterday. Would love some more feedback. Also anything I may be missing to get this ready to merge.

> Re: Bolding default option for windows

Hmm... [doesn't look](https://github.com/electron/electron/issues/3908) like this is currently possible in electron.

File: app/src/lib/editors/launch.ts [hunk 1]
```diff
@@ -32,10 +35,45 @@ export async function launchExternalEditor(
...
+const editorOpenToLineCommands: { [key: string]: any } = {
+  'Visual Studio Code': async (
+    editor: FoundEditor,
+    fullPath: string,
+    line: string
+  ) => {
+    const execName = 'code'
+    if (!(await executableExists(execName))) {
+      throw new Error(
+        `Could not find the '${execName}' executable for '${editor.editor}' on path. Please install and try again.`
```
Comment (digitalmaster) [inline app/src/lib/editors/launch.ts]: NIT: Maybe add link to instructions for installing the vscode cli?

File: app/src/lib/editors/launch.ts [hunk 2]
```diff
@@ -32,10 +35,45 @@ export async function launchExternalEditor(
   if (editor.usesShell) {
     spawn(`"${editorPath}"`, [`"${fullPath}"`], { ...opts, shell: true })
   } else if (__DARWIN__) {
```
Comment (digitalmaster) [inline app/src/lib/editors/launch.ts]: NIT: Because of this guard diff double clicks will not go to the specified line.

**Review by digitalmaster — COMMENTED** [review]
Self review I.

**Review by sergiou87 — CHANGES_REQUESTED** [review]
Thanks for your contribution! 💖 

I tried this out and noticed that double-clicking files from the list of files changed inside of a commit doesn't work as intended.

The scenario I have is this:
1. Open VS Code with the `desktop/desktop` repo
2. Open GitHub Desktop with a different repo selected (let's say `sergiou87/test-repo`), a repo with `README.md`
3. Look for a commit that changed `README.md`
4. Double click on `README.md` inside of such commit

Instead of opening `README.md` of `sergiou87/test-repo`, it opens `README.md` in the current instance of VS Code with `desktop/desktop` 🤔 

Aside from that bug, I just noticed you also added "double-click on diff line to open file at that line", which I _think_ is something we didn't agree on? I will check with the team next week and get back to you 😄

**Comment by digitalmaster** [comment]
Thanks for taking a look 🙌🏽

> Instead of opening README.md of sergiou87/test-repo, it opens README.md in the current instance of VS Code with desktop/desktop 🤔

Excellent find!! 🥳 I was able to reproduce it. Looks like I need to go through all the code paths and ensure that we're always generating a full path instead of a relative path everywhere. 

Gonna also see if I can refactor this so that this happens consistently in one place.

> Aside from that bug, I just noticed you also added "double-click on diff line to open file at that line", which I think is something we didn't agree on? I will check with the team next week and get back to you 😄

Oh I see. My apologies. I'm not familiar with what specifically was agreed upon or how things like this work on your team but boy do I hope this is allowed. To be fair tho the only place this was mentioned was [here](https://github.com/desktop/desktop/issues/2620#issuecomment-1318914036):
CleanShot 2023-05-05 at 11 53 29@2x


Any reason you think the team might ok with double click to a file but not to the specific line number? I've been using it while developing this feature and I really think users will love this. It's so common that i'd wanna make a change to a specific line while reviewing a diff.. this makes that so fast. Please please please 🤞🏽🙏🏽 😭 😅

Thanks again for taking the time to review <3

**Comment by sergiou87** [comment]
I'm sorry but we just chatted and are not convinced about it 😕 As mentioned: it's not very discoverable and kind of conflicts with the existing interactions. Adding something like that would require more thought and discussion, but our current priorities don't give us much time for it 😞 

We think it's better to focus on "double-click on file" in this PR, and anything beyond that should be discussed that in a new issue. But again, we don't have the bandwidth to engage in such discussion right now, so it might take a while until we take a decision 😓

**Comment by digitalmaster** [comment]
Got it.. i'll open a new issue. I just finished reverting the feature. Take a look when you have a moment 🙏🏽

File: app/src/ui/changes/changes-list.tsx [hunk 3]
```diff
@@ -151,6 +151,14 @@ interface IChangesListProps {
    * @param path The path of the file relative to the root of the repository
    */
   readonly onOpenItem: (path: string) => void
+
+  /**
+   * Called to open a file it its the default external editor
```
Comment (sergiou87) [inline app/src/ui/changes/changes-list.tsx]: Nit:
```suggestion
   * Called to open a file it the default external editor
```

File: app/src/ui/history/selected-commits.tsx [hunk 4]
```diff
@@ -116,6 +116,17 @@ export class SelectedCommits extends React.Component<
     this.props.dispatcher.changeFileSelection(this.props.repository, file)
   }
 
+  private onRowDoubleClick = (row: number) => {
+    const files = this.props.changesetData.files
+    const file = files[row]
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
Comment (sergiou87) [inline app/src/ui/history/selected-commits.tsx]: This should be either `log.error(…)` or nothing. I think we can just remove it:
```suggestion
```

File: app/src/ui/open-pull-request/pull-request-files-changed.tsx [hunk 5]
```diff
@@ -224,6 +231,17 @@ export class PullRequestFilesChanged extends React.Component<
     )
   }
 
+  private onRowDoubleClick = (row: number) => {
+    const files = this.props.files
+    const file = files[row]
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
Comment (sergiou87) [inline app/src/ui/open-pull-request/pull-request-files-changed.tsx]: Same:
```suggestion
```

File: app/src/ui/stashing/stash-diff-viewer.tsx [hunk 6]
```diff
@@ -64,11 +71,26 @@ export class StashDiffViewer extends React.PureComponent<IStashDiffViewerProps>
   private onSelectedFileChanged = (file: CommittedFileChange) =>
     this.props.dispatcher.selectStashedFile(this.props.repository, file)
 
+  private onRowDoubleClick = (row: number) => {
+    const files = this.getFiles()
+    const file = files[row]
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
Comment (sergiou87) [inline app/src/ui/stashing/stash-diff-viewer.tsx]: Same:
```suggestion
```

File: app/src/ui/changes/sidebar.tsx [hunk 7]
```diff
@@ -249,6 +249,26 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
...
+  private onOpenItemInExternalEditor = (path: string) => {
+    this.props.onOpenInExternalEditor(path)
+  }
+
+  private onRowDoubleClick = (row: number) => {
+    const workingDirectory = this.props.changes.workingDirectory
+    const file = workingDirectory.files[row]
+
+    if (!file) {
+      console.error('double click on row with no file - what?')
```
Comment (sergiou87) [inline app/src/ui/changes/sidebar.tsx]: Same:
```suggestion
```

File: app/src/ui/changes/sidebar.tsx [hunk 8]
```diff
@@ -249,6 +249,26 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
     const fullPath = Path.join(this.props.repository.path, path)
     openFile(fullPath, this.props.dispatcher)
   }
+  /**
+   * Called to open a file it its the default external editor
```
Comment (sergiou87) [inline app/src/ui/changes/sidebar.tsx]: Nit:
```suggestion
   * Called to open a file it the default external editor
```

File: app/src/ui/changes/changes-list.tsx [hunk 9]
```diff
@@ -981,6 +982,7 @@ export class ChangesList extends React.Component<
               isCommitting: isCommitting,
             }}
             onRowClick={this.props.onRowClick}
+            onRowDoubleClick={this.props.onRowDoubleClick}
```
Comment (sergiou87) [inline app/src/ui/changes/changes-list.tsx]: Just curious: why bubbling this up instead of also calling `onOpenItemInExternalEditor`?

**Review by sergiou87 — CHANGES_REQUESTED** [review]
It looks good overall! Thank you for all those changes 🙇‍♂️ 
I found a few nits/typos, some `console.blahblah` that should be removed, but other than that looks good. 

I also wonder why in some controls you have bubbled up both double click and "open in external editor" instead of grouping both in the latter 🤔

File: app/src/ui/history/selected-commits.tsx (hunk 4 above)
Comment (digitalmaster) [reply app/src/ui/history/selected-commits.tsx]: Yea, agreed this is unnecessary (was just blindly following a pattern i'd seen 🙈) 👍🏽

File: app/src/ui/changes/changes-list.tsx (hunk 9 above)
Comment (digitalmaster) [reply app/src/ui/changes/changes-list.tsx]: Ah good catch 🥳 -  no need to hoist this any further 👍🏽

**Comment by digitalmaster** [comment]
Thanks for the feedback. Iterated. Give it another pass when you have a chance 🙏🏽

File: app/src/ui/history/selected-commits.tsx
Comment (sergiou87) [inline app/src/ui/history/selected-commits.tsx:1]: Line 400 needs to change from
```typescript
action: () => this.props.onOpenInExternalEditor(fullPath),
```
to
```typescript
action: () => this.props.onOpenInExternalEditor(file.path),
```

File: app/src/ui/changes/sidebar.tsx [hunk 10]
```diff
@@ -249,6 +249,14 @@ export class ChangesSidebar extends React.Component<IChangesSidebarProps, {}> {
     const fullPath = Path.join(this.props.repository.path, path)
     openFile(fullPath, this.props.dispatcher)
   }
+  /**
+   * Called to open a file in its the default external editor
```
Comment (sergiou87) [inline app/src/ui/changes/sidebar.tsx]: I noticed you left part of the typo here:
```suggestion
   * Called to open a file in the default external editor
```

**Review by sergiou87 — CHANGES_REQUESTED** [review]
Thank you for those quick changes! 💖 

Unfortunately, I tested it and noticed that going to a commit -> right click on a file -> Open in _Whatever_ doesn't work anymore, because clicking on that still uses the full path instead of the relative one.

I think that should be it 🤞

File: app/src/ui/history/selected-commits.tsx
Comment (digitalmaster) [reply app/src/ui/history/selected-commits.tsx:1]: Oh that's right... Nice catch! Thanks for testing 🙌🏽

File: app/src/ui/changes/sidebar.tsx (hunk 10 above)
Comment (digitalmaster) [reply app/src/ui/changes/sidebar.tsx]: ha! You're patience knows no bounds.. I'll get it right eventually 😆 🤦🏽‍♂️

Comment (sergiou87) [reply app/src/ui/changes/sidebar.tsx]: LOL no worries 🫂 I make those mistakes (and much worse) all the time 😂

**Review by sergiou87 — APPROVED** [review]
Code looks great and works like a charm! Thank you for contribution and your patience through all this process 🙇‍♂️🙇‍♂️🙇‍♂️🙇‍♂️

**Comment by digitalmaster** [comment]
It was my pleasure. Thank you for the time to review and get this ready for primetime 🙌🏽

**Comment by mr21** [comment]
Hi great feature :) thanks you for that

But currently this feature works also when we double click on the checkbox:
image

And we can find a reason to double click on those checkbox to reset the particular line by line selection

**Comment by sergiou87** [comment]
Hey @mr21 👋 

Thanks for reporting that, I missed that comment and just ran into it 😱 We will take a look at it. It's not something I've suffered myself but I understand how it could be super annoying.

**EDIT:** Fixed in #17229

**Comment by mr21** [comment]
Cool, it's working perfectly now 🥳

**Comment by digitalmaster** [comment]
This was removed in the latest version. Are there any plans to add this back?

**Comment by tidy-dev** [comment]
@digitalmaster What version are seeing it not working? Just tried on the latest beta and it worked as expected.

**Comment by digitalmaster** [comment]
Oh sorry i'm talking about click to go directly to specific lines. I forgot that you guys didn't think it was worth adding that part (reverted and made [separate issue](https://github.com/desktop/desktop/issues/20254)) so I kept my own build with the feature and have been using it for years now.

That said, are you sure you guys didn't add support for this afterwards? IIRC at some point I stopped having to rollback  to my personal build after every update?

**Comment by tidy-dev** [comment]
Ah. I could be wrong, but I do not believe we added support for that.

**Comment by digitalmaster** [comment]
Ok good to know there's a small chance i'm not completely making that up 😆.

For me, this is still the most use action when i'm reviewing a diff - I'm going through my changes and I see a line that want to tweak, I double click to go right to it (no need to go to file then scroll or search).

**Comment by digitalmaster** [comment]
@tidy-dev Would the team open to a PR that adds this feature back? If so I could make some time to do this again in the coming weeks.

**Comment by MarkHaakman** [comment]
@digitalmaster If the main GitHub team still does not want to add support for double clicking lines to open it in an editor (which is in their right of course), you could also open a pull request to https://github.com/pol-rivero/github-desktop-plus , which is an actively maintained fork with additional features. 

Getting this functionality in the main repo is of course far nicer, but if the team decides against it, this is the next best alternative I guess.

**Pull request merged**